### ISSUE TITLE
SFDX CLI - multiple sfdx force:mdapi:deploy commands competing for the same folder/file? 

### ISSUE THREAD

**Issue by neowit** [forcedotcom/cli#1299]
### Summary

Multiple `sfdx force:mdapi:deploy` commands executed in parallel from the same MDAPI folder but to different Orgs conflict with each other

This issue manifests itself in all sfdx cli versions starting 7.82.0 (did not test in earlier versions) and up to (and including) current. 

### Description:

As part of CI deployment the same PR needs to be validated against several salesforce orgs.
To do that a CI script has the following (not exact copy, but to give an idea)

```
SF_DEPLOYMENT_DIR="/some/path/here" # same deploydir is re-used for all orgs in deployment

# org 1
sfdx force:mdapi:deploy --checkonly --json --targetusername deploymentuser@org1.com \
            --deploydir "$SF_DEPLOYMENT_DIR" --testlevel $TEST_LEVEL --runtests "$TESTS_TO_RUN1" & # run in background, async

#org 2
sfdx force:mdapi:deploy --checkonly --json --targetusername deploymentuser@org2.com \
            --deploydir "$SF_DEPLOYMENT_DIR" --testlevel $TEST_LEVEL --runtests "$TESTS_TO_RUN2" & # run in background, async

#org 3	
sfdx force:mdapi:deploy --checkonly --json --targetusername deploymentuser@org3.com \
            --deploydir "$SF_DEPLOYMENT_DIR" --testlevel $TEST_LEVEL --runtests "$TESTS_TO_RUN3" & # run in background, async
```			
			
The issue is - if commands for `org 2` or `org 3` are executed just like it is above (i.e. at the same time) then most of the time at least 1 (more often 2) deployments fails with either `UNKNOWN_ERROR` or `'package.xml' missing` error, or some other error related to incomplete metadata folder/file structure
It looks as if each subsequent `sfdx force:mdapi:deploy` has a clash (at file system level) with preceding `sfdx force:mdapi:deploy`

If we add a `sleep 120` after each deployment, i.e. give each  `sfdx force:mdapi:deploy` command some time to start in isolation, then the whole thing almost never fails.

This version works (but adds undesirable pause):

```
SF_DEPLOYMENT_DIR="/some/path/here" # same deploydir is re-used for all orgs in deployment

# org 1
sfdx force:mdapi:deploy --checkonly --json --targetusername deploymentuser@org1.com \
            --deploydir "$SF_DEPLOYMENT_DIR" --testlevel $TEST_LEVEL --runtests "$TESTS_TO_RUN1"  & # run in background, async
sleep 120

#org2
sfdx force:mdapi:deploy --checkonly --json --targetusername deploymentuser@org2.com \
            --deploydir "$SF_DEPLOYMENT_DIR" --testlevel $TEST_LEVEL --runtests "$TESTS_TO_RUN2"  & # run in background, async
sleep 120

#org3	
sfdx force:mdapi:deploy --checkonly --json --targetusername deploymentuser@org.com \
            --deploydir "$SF_DEPLOYMENT_DIR" --testlevel $TEST_LEVEL --runtests "$TESTS_TO_RUN3"  & # run in background, async
sleep 120 # not sure if this one make any difference, but that's what currently working
```

The problem is - PR validation is done many times each day by many developers and `(N-1) x 2`  additional minutes (where N is number of orgs in each validation) spent to start deployments in every PR validation quickly adds up.

**Question** - what can be done to stop multiple `sfdx force:mdapi:deploy` executed **in parallel** from conflicting with each other?
Thank you

PS: this problem looks like it is caused because some (writeable) file system resource is being re-used when multiple `sfdx force:mdapi:deploy` are executed on the same machine at the same time. If each `sfdx force:mdapi:deploy` was truly resource independent (for stuff it writes) then this would probably solve this `clash` issue

**Comment by AllanOricil** [comment]
@neowit Instead of having one job doing N org validations, make N jobs, each with its own copy of the repo, to run a validation. This way you won't need to repeat/write the sfdx command N times. It also follows a DRY implementation and your job will be reusable.

**Comment by neowit** [comment]
Hello @AllanOricil 
Thank you for the suggestion. 
As I wrote in my message - the above is an example to give an idea. N is not a fixed number and there is no repetition in the real CI script. Each deployment is indeed its own job which re-use metadata prepared by preceding stages of deployment.
Those 3 `sfdx force:mdapi:deploy` lines in my example are just a simplified way to illustrate what's going on.
Having to prepare N versions of same `SF_DEPLOYMENT_DIR` with identical content does not feel like a DRY implementation.

**Comment by AllanOricil** [comment]
You said that the problem happens if you reuse the same folder to run the validation command. If you are indeed using different jobs, how is it possible you are still facing this problem?
You are probably not checking out the repo again in every job. Every job should work with its own copy of the repo. Build a job that can be reused, like a function. Then you spawn multiple instances of this job to avoid repeating sfdx commands several times.

This is an example of implementation I did using Github Actions for a customer that works well:

````yml
name: Validation Workflow

on:
    pull_request:
        types: [opened, synchronize, ready_for_review]
        branches:
            - main

    workflow_dispatch:

jobs:

    validate-vlocdevbas:
        uses: VodafoneIS/sf-metadata/.github/workflows/validation.yml@main
        with:
            ENVIRONMENT: vlocdevbas
        secrets:
            SF_ORG_ALIAS: ${{ secrets.SF_ORG_ALIAS }}
            SF_INSTANCE_URL: ${{ secrets.SF_INSTANCE_URL }}
            SF_SFDX_AUTH_URL: ${{ secrets.SF_SFDX_AUTH_URL }}
            TAG_NAME: ${{ secrets.TAG_NAME }}
    
    validate-vlocdevpro:
        uses: VodafoneIS/sf-metadata/.github/workflows/validation.yml@main
        with:
            ENVIRONMENT: vlocdevpro
        secrets:
            SF_ORG_ALIAS: ${{ secrets.SF_ORG_ALIAS }}
            SF_INSTANCE_URL: ${{ secrets.SF_INSTANCE_URL }}
            SF_SFDX_AUTH_URL: ${{ secrets.SF_SFDX_AUTH_URL }}
            TAG_NAME: ${{ secrets.TAG_NAME }}
````

In this implementation I have used a feature called "Reusable Workflow" that was created based on DRY implementations. Instead of repeating a sequence of jobs multiple times, I separate them in diferent templates, like functions, and then I reuse them. This way I can avoid repeating cli commands. 
Each validation workflow does the following:

- check out the repo. This way each validation works with its own copy of the repo and this error you mentioned won't happen.
- authenticate to an org
- compute metadata diff using sgd plugin
- run sfdx validation command. If it failed exit with 1.
- compute code coverage. If code coverage is below MIN_COVERAGE(default 75) exit with 1.

**Comment by AllanOricil** [comment]
My suggestion would solve the problem, but the bug would still be there. I think it is related to the folder being locked by the OS. But since the error is UNKNOWN, this is just a guess.

**Comment by neowit** [comment]
Hello @AllanOricil 
Thank you for taking the time to write an example.

> If you are endeed using different jobs, how are you having the problem?

I am sorry if my explanation was unclear/confusing. I did not mean to imply that the pipeline in question is on github. It is not  and its internals are not relevant to the issue. I can reproduce the error on my local machine without any pipelines just as well. 

When I wrote `job` I meant OS level process (shell script executed in background, asynchronously) not `github action` job.
So if there are N deployments then N separate threads will be spawn (not several pipeline jobs) on the deployment agent. 

> You are probably not checking out the repo again in every job. 

There is no need to checkout the same source multiple times in this instance. If there was a need to provide separate source metadata folder for each spawned deployment then `sleep 120` (which cures the problem in my case) would not make any difference.

`UNKNOWN_ERROR` is returned by salesforce in response to `sfdx force:mdapi:deploy` call. Sometimes it has additional description and reports missing files explicitly which lead me to believe that multiple/parallel `sfdx force:mdapi:deploy` with the same source metadata folder compete with each other for the same OS resource.

What I was hoping to achieve by starting this issue is one of two things:
**A**. someone could point me to an undocumented switch in `sfdx force:mdapi:deploy` command line which stops it from re-using the same file system resource when compressing metadata folder before it is sent to salesforce 
OR
**B**. someone who is capable of fixing `sfdx force:mdapi:deploy` adds it to their backlog so it gets fixed eventually and we can remove that artificial delay from CI deployment script

And super bonus would be - if there was a switch which tells next `sfdx force:mdapi:deploy` that it can re-use deployment package prepared by previous `sfdx force:mdapi:deploy` so it does not even need to waste any time for checking & compressing metadata before it is sent to another Org.

**Comment by AllanOricil** [comment]
@neowit from you comments, I understood you still want to use multiple sfdx deploy commands sequentially (I know they are in different processes and running in parallel), and use the same folder.
Do you really need it to be like that? If yes,  could you show why it must be done like that? 
what limits you from changing it knowing it currently does not work without the sleep hack?
I still don't understand why, for your use case, you want to keep using the same folder. Could you explain that?

The example I showed using github was to show to you a way that works and avoids the problem you have. I never assumed you were using Gtihub. It was there to better explain the suggestion. The way you are doing running multiple sfdx commands in the same folder and adding the sleep hack is not a good way of handling it. Instead of having one process starting the verification to multiple orgs, you could actually trigger multiple processes, one for each org. Each process would have its own copy of the repo and each would start the verification. This way you won't need the sleep hack. And the solution will be much cleaner. This is what I was trying to explain with the github action example.

I have worked with multiple pipeline implementations already: Bitbucket Pipelines, Jenkins and now Github Actions.  In all these implementations we spawn multiple Jobs, one for each environment, and each containing its own copy of the repository. We never shared the same folder, or the ran a sfdx command to multiple orgs in the same process. Each job handles its own org.

**Comment by AllanOricil** [comment]
> @neowit from you comments, I understood you still want to use multiple sfdx deploy commands sequentially (I know they are in different processes and running in parallel), and use the same folder. Do you really need it to be like that? If yes, could you show why it must be done like that? what limits you from changing it knowing it currently does not work without the sleep hack? I still don't understand why, for your use case, you want to keep using the same folder. Could you explain that?
> 
> The example I showed using github was to show to you a way that works and avoids the problem you have. I never assumed you were using Gtihub. It was there to better explain the suggestion. The way you are doing running multiple sfdx commands in the same folder and adding the sleep hack is not a good way of handling it. Instead of having one process starting the verification to multiple orgs, you could actually trigger multiple processes, one for each org. Each process would have its own copy of the repo and each would start the verification. This way you won't need the sleep hack. And the solution will be much cleaner. This is what I was trying to explain with the github action example.
> 
> I have worked with multiple pipeline implementations already: Bitbucket Pipelines, Jenkins and now Github Actions. In all these implementations we spawn multiple Jobs, one for each environment, and each containing its own copy of the repository. We never shared the same folder or even ran multiple sfdx commands to multiple orgs in the main process. Each job handles its own environment.

**Comment by neowit** [comment]
Hello @AllanOricil 
The approach with multiple jobs is going to be significantly more expensive in my case because each parallel job cost extra.
We do use parallel jobs in other instances, where appropriate, but not in this case because that would consume too many parallel jobs unnecessarily.  

>  The way you are doing running multiple sfdx commands in the same folder and adding the sleep hack is not a good way of handling it.

Indeed, the sleep hack is not a good way, that's why I opened this issue.
On the other hand - running multiple deployments (re-using the same source and all of the work done to prepare that source) **is** a good way. 
There are several steps (done earlier in the job tasks) that are required to convert and prepare the source from version control repo to make it deployable with `sfdx force:mdapi:deploy`. i.e. it is not just `git checkout branch + sfdx force:mdapi:deploy`.

Anyway - your questions prompted me to check what will happen if each deployment command was given its own folder (by copying the original `$SF_DEPLOYMENT_DIR` content into a temporary folder and feeding each `sfdx force:mdapi:deploy` its own clone of the original `$SF_DEPLOYMENT_DIR`. Have not had a chance to run many tests yet, but so far it seems working. Still a hack, and a workaround, but at least we do not need to depend on an arbitrary pause. Will need to run more tests to confirm if it is really working every time...

Have also tried to create a symlink to the original `SF_DEPLOYMENT_DIR` (instead of doing complete folder copy) and feeding that symlink to  `sfdx force:mdapi:deploy` but it would appear that `sfdx force:mdapi:deploy` does not support symlinks and simply treats such source as empty and starts with running tests straight away.

**Comment by mshanemc** [comment]
I think the upcoming update of `force:mdapi:deploy` is going to solve this. @W-9603711@

I'm going to put a note about this as a feature for that work to make sure simultaneous deploys work.

**Comment by uip-robot-zz** [comment]
This issue has been linked to a new work item: [W-10261007](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000UskKjYAJ/view)

**Comment by cristiand391** [comment]
Hi @neowit 👋🏼 , I've been trying to reproduce this using [dreamhouse-lwc](https://github.com/trailheadapps/dreamhouse-lwc) and this script:

```bash
#! /usr/bin/env bash

sfdx force:mdapi:deploy --checkonly --deploydir mdapi_output -u dreamhouse1 > deploy1.log 2>&1 &
sfdx force:mdapi:deploy --checkonly --deploydir mdapi_output -u dreamhouse2 > deploy2.log 2>&1 &
sfdx force:mdapi:deploy --checkonly --deploydir mdapi_output -u dreamhouse3 > deploy3.log 2>&1 &
sfdx force:mdapi:deploy --checkonly --deploydir mdapi_output -u dreamhouse4 > deploy4.log 2>&1 &
sfdx force:mdapi:deploy --checkonly --deploydir mdapi_output -u dreamhouse5 > deploy5.log 2>&1 &
sfdx force:mdapi:deploy --checkonly --deploydir mdapi_output -u dreamhouse6 > deploy6.log 2>&1 &

wait
echo 'Done'
```
but didn't get any error locally.

Some weeks ago we release two new mdapi beta commands that use the [SDR](https://github.com/forcedotcom/source-deploy-retrieve/) library. This library does all the work in memory so you shouldn't get any FS conflict.

Could you try switching to `force:mdapi:beta:deploy` and tell us how it goes? These commands are available in [sfdx ≤ v7.133.0](https://github.com/forcedotcom/cli/tree/main/releasenotes/sfdx#71330-jan-13-2022)

**Comment by neowit** [comment]
Hello @cristiand391 

Thank you for looking into this and your advice regarding the version to use.

I have re-tested today and happy to report that I am unable to reproduce the error with `sfdx-cli@7.134.0`

Here is how I tested (same deployment commands & metadata source used in each test):
1. installed `sfdx-cli@7.134.0` - result: no error
1. installed `sfdx-cli@7.122.1` - result: sf:UNKNOWN_EXCEPTION error, deployments failed to start (as expected)
3. installed `sfdx-cli@7.134.0` again - result: no error

FYI - in my successful test I used `sfdx force:mdapi:deploy` not `sfdx force:mdapi:beta:deploy` as you suggested because that's what I had to use with  `sfdx-cli@7.122.1`. Not sure if that makes much difference but thought I should mention.

I believe this issue can be closed as resolved.
Thank you everyone who invested their time and effort in this.
